src/Microsoft.VisualStudio.ProjectSystem.Managed.VS.UnitTests/ProjectSystem/VS/Tree/Dependencies/Snapshot/TargetedDependenciesSnapshotTests.cs

```diff
@@ -573,6 +573,47 @@ public void TFromChanges_RemovedAndAddedChanges()
             Assert.True(snapshot.DependenciesWorld.ContainsKey(@"tfm1\xxx\addeddependency3"));
         }
 
+        /// <summary>
+        /// Added because circular dependencies can cause stack overflows
+        /// https://github.com/dotnet/project-system/pull/3474
+        /// </summary>
+        [Fact]
+        public void TCheckForUnresolvedDependencies_CircularDependency_DoesNotRecurseInfinitely()
+        {
+            var dependencyModelTop1 = IDependencyFactory.FromJson(@"
+            {
+                ""ProviderType"": ""Xxx"",
+                ""Id"": ""tfm1\\xxx\\topdependency1"",
+                ""Name"":""TopDependency1"",
+                ""Caption"":""TopDependency1"",
+                ""SchemaItemType"":""Xxx"",
+                ""Resolved"":""true"",
+                ""DependencyIDs"": [ ""tfm1\\xxx\\topdependency2"" ]
+            }", icon: KnownMonikers.Uninstall, expandedIcon: KnownMonikers.Uninstall);
+
+            var dependencyModelTop2 = IDependencyFactory.FromJson(@"
+            {
+                ""ProviderType"": ""Xxx"",
+                ""Id"": ""tfm1\\xxx\\topdependency2"",
+                ""Name"":""TopDependency2"",
+                ""Caption"":""TopDependency2"",
+                ""SchemaItemType"":""Xxx"",
+                ""Resolved"":""true"",
+                ""DependencyIDs"": [ ""tfm1\\xxx\\topdependency1"" ]
+            }", icon: KnownMonikers.Uninstall, expandedIcon: KnownMonikers.Uninstall);
+
+            var previousSnapshot = new TargetedDependenciesSnapshot(
+                dependenciesWorld: new Dictionary<string, IDependency>()
+                {
+                    { dependencyModelTop1.Id, dependencyModelTop1 },
+                    { dependencyModelTop2.Id, dependencyModelTop2 },
+                },
+                topLevelDependencies: new List<IDependency>() { dependencyModelTop1 });
+
+            // verify it doesn't stack overflow
+            previousSnapshot.CheckForUnresolvedDependencies(dependencyModelTop1);   
```

Comment:
Have you verified that this test fails without your fix?

Reply:
Yes, it fails without the fix.